src/compiler/checker.ts

```diff
@@ -7847,6 +7848,28 @@ namespace ts {
                 return t.immediateBaseConstraint;
             }
 
+            function checkMaximumConstraintDepthExceeded(t: Type) {
+                return constraintDepth >= 50 || (t.flags & TypeFlags.Conditional && (perInstanceDepth.get("" + getTypeId(getNodeLinks((t as ConditionalType).root.node).resolvedType!)) || 0) >= 10);
+            }
+
+            function incrementConstraintDepth(t: Type) {
+                constraintDepth++;
+                if (t.flags & TypeFlags.Conditional) {
+                    const base = getNodeLinks((t as ConditionalType).root.node).resolvedType!;
+                    const id = "" + getTypeId(base);
+                    perInstanceDepth.set(id, (perInstanceDepth.get(id) || 0) + 1);

```

Comment:
Why all the extra indirections here? Couldn't you just store the counter as a property in the resolved type?